art/attacks/evasion/carlini.py

```diff
@@ -797,3 +797,563 @@ def _check_params(self) -> None:
+                    logger.debug("Number of improved L0 distances: %i", int(np.sum(improved_adv)))
+                    if np.sum(improved_adv) > 0:
+                        best_l0dist_batch[improved_adv] = l0dist[improved_adv]
+                        best_x_adv_batch[improved_adv] = x_adv_batch[improved_adv]
+
+                    c_double[overall_attack_success] = False
+                    c_current[overall_attack_success] = (c_lower_bound + c_current)[overall_attack_success] / 2
+
+                    c_old = c_current
+                    c_current[~overall_attack_success & c_double] *= 2
+
+                    c_current1 = (c_current - c_lower_bound)[~overall_attack_success & ~c_double]
+                    c_current[~overall_attack_success & ~c_double] += c_current1 / 2
+                    c_lower_bound[~overall_attack_success] = c_old[~overall_attack_success]
+
+                c_final[batch_index_1:batch_index_2] = c_current
+                x_adv[batch_index_1:batch_index_2] = best_x_adv_batch
+
+            logger.info(
+                "Success rate of C&W L_2 attack: %.2f%%",
+                100 * compute_success(self.estimator, x, y, x_adv, self.targeted, batch_size=self.batch_size),
+            )
+
+            # If the L_2 attack can't find any adversarial examples with the new activation, return the last one
+            z_logits, l2dist, loss = self._loss(x, x_adv, y, c_final)
+            attack_success = loss - l2dist <= 0
+            # l0dist = np.sum((np.abs(x_batch - x_adv_batch) > self._perturbation_threshold).astype(int), axis=1)
+            l0dist = np.sum((np.abs(x - x_adv) > self._perturbation_threshold).astype(int), axis=1)
+            improved_adv = attack_success & (l0dist < best_l0dist)
+            if np.sum(improved_adv) > 0:
+                final_adversarial_example[improved_adv] = x_adv[improved_adv]
+            else:
+                return x * (old_activation == 0).astype(int) + final_adversarial_example * old_activation
+
+            # Compute the gradients of the objective function evaluated at the adversarial instance
+            x_adv_tanh = original_to_tanh(x_adv, clip_min, clip_max, self._tanh_smoother)
+            objective_loss_gradient = -self._loss_gradient(
+                z_logits,
+                y,
+                x,
+                x_adv,
+                x_adv_tanh,
+                c_final,
+                clip_min,
+                clip_max,
+            )
+            perturbation_L1_norm = np.abs(x_adv - x)
+
+            # gradient * perturbation tells how much reduction to the objective function we obtain for each attribute
+            objective_reduction = np.abs(objective_loss_gradient) * perturbation_L1_norm
+
+            # Put a huge nomber as objective_reduction value for fixed feature (in order not to select them again)
+            # was np.inf before, but inf * 0 = nan
+            objective_reduction += 999999999999999 * (activation == 0).astype(int)
```

Comment:
I'd suggest to find a way for adding `np.inf` (e.g. using `np.where`) instead of resorting to hard-coded constant, if possible.

Reply:
Good idea, `np.where()` does the job perfectly, thank you.